src/storage-area.cpp

```diff
@@ -0,0 +1,202 @@
+            path = fs::path(storage_root)
+                    .append("/by-uuid/")
+                    .append(b1)
+                    .append(b2)
+                    .append(uuid)
+                    .append(".DCM");
+            break;
+        case OrthancPluginContentType_DicomAsJson:
+            path = fs::path(storage_root)
+                    .append("/json/")
+                    .append(b1)
+                    .append(b2)
+                    .append(uuid)
+                    .append(".json");
+            break;
+        case OrthancPluginContentType_DicomUntilPixelData:
+            path = fs::path(storage_root)
+                    .append("/no-pixel/")
+                    .append(b1)
+                    .append(b2)
+                    .append(uuid)
+                    .append(".DCM");
+            break;
+        case _OrthancPluginContentType_INTERNAL:
+            path = fs::path(storage_root)
+                    .append("/internal/")
+                    .append(b1)
+                    .append(b2)
+                    .append(uuid);
+            break;
+        case OrthancPluginContentType_Unknown:
+            path = fs::path(storage_root)
+                    .append("/unknown-files/")
+                    .append(b1)
+                    .append(b2)
+                    .append(uuid);
+            break;
+    }
+    return path;
+}
+
+OrthancPluginErrorCode WriteDicomFile(DicomFile dicom, const char *uuid){
+    const fs::path storage_root(globals::storage_location);
+    std::unique_ptr<char[]> content = nullptr;
+    size_t size = 0;
+    if(dicom.IsValid()) {
+        DBInterface::HandlePHI(dicom);
+        auto filtered = dicom.ApplyFilter(globals::filter_list);
+        if (std::get<0>(filtered)) {
+            content = std::move(std::get<1>(filtered));
+            size = std::get<2>(filtered);
+            if(size == 0){
+                // todo: probably a better error code
+                return OrthancPluginErrorCode_EmptyRequest;
+            }
+        }
+        // write to disk
+        // todo: add uuid directory portion
+        fs::path master_path = fs::path(storage_root)
+                .append("/by-uuid/")
+                .append(uuid)
+                .append(".DCM");
+        fs::create_directories(master_path);
+        std::fstream file(master_path, std::ios::binary | std::ios::out);
+        file.write(content.get(),size);
+        file.close();
+        // set file permissions
+        // todo: permission debug info?
+        fs::file_status master_status = fs::status(master_path);
+        master_status.permissions(globals::file_permissions);
+
+        // create hard links
+        auto hardlink_to = [&](std::string groupby, std::string group) {
+            fs::path link = fs::path(storage_root)
+                    .append(groupby)
+                    .append(group)
+                    .append(uuid)
+                    .append(".DCM");
```

Comment:
Perhaps the hard links should use buckets as well? maybe just one bucket if any?